Refactor the CTLog config code, add multi-fulcio support.

vaikas opened the pull request:
Signed-off-by: Ville Aikas <vaikas@chainguard.dev>

<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Pleases use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
Addresses (does not fix): https://github.com/sigstore/fulcio/issues/589

Start refactoring the code that manages the CTLog configuration and add support for adding/removing trusted
Fulcio certs + make it testable :)

This one currently adds a library, that I then intend to use, but since I have a question for @haydentherapper wanted to
share earlier.

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+// technically they are not part of the config, however because we create a
+// secret/CM that we then mount, they need to be synced.
+type CTLogConfig struct {
+	PrivKey         *rsa.PrivateKey
```
Can we use an ECDSA key instead? It'll produce smaller signatures over SCTs, and that will result in smaller certificates since SCTs are embedded in issued certs.

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+	// FulcioCerts contains one or more Root certificates for Fulcio.
+	// It may contain more than one if Fulcio key is rotated for example, so
+	// there will be a period of time when we allow both.
+	FulcioCerts [][]byte
```
nit on the comment, it might also contain multiple if we choose to support admitting certificates from fulcio instances run by others.

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+	if pubPEM == nil {
+		return nil, fmt.Errorf("did not find valid public PEM data")
+	}
+	pubKey, err := x509.ParsePKCS1PublicKey(pubPEM.Bytes)
```
when switching from rsa to ecdsa, this'll need to change to ParsePKIXPublicKey

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+		return nil, fmt.Errorf("failed to decrypt private PEMKeyFile: %w", err)
+	}
+	ret.PrivKeyPassword = pb.Password
+	privKey, err := x509.ParsePKCS1PrivateKey(privatePEMBlock)
```
This will need to change to ParsePKCS8PrivateKey

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+// multiple.
+func (c *CTLogConfig) marshalSecrets() (map[string][]byte, error) {
+	// Encode private key to PKCS#1 ASN.1 PEM.
+	block := &pem.Block{
```
change marshalling to "PRIVATE KEY" and MarshalPkcs8PrivateKey, and below change to "PUBLIC KEY" and MarshalPKIXPublicKey

haydentherapper commented on pkg/ctlog/config.go:
```diff
@@ -0,0 +1,324 @@
+		fulcioKey := fmt.Sprintf("fulcio-%d", i)
+
+		// Fetch only root certificate from the chain
+		// TODO(Hayden): Is this what we should be adding, or the intermediates,
```
This looks correct, it should only be the last certificate in the chain, the root.

haydentherapper commented on pkg/ctlog/config_test.go:
```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate serial Number: %w", err)
+	}
+	rootCA := &x509.Certificate{
```
See https://github.com/sigstore/rekor/blob/main/pkg/pki/x509/testutils/cert_test_utils.go, I created some test utils for creating root certs that are more representative of Sigstore root certs.

haydentherapper reviewed (commented):
This is looking great!

One higher level question - Since we'll be designing this to support multiple CT log instances, are there any changes we should be making in this refactor now to start supporting that? For example, if there's any places that expect one CT log config, making it an array? From looking over this, I think no if this is just setting up a single CT log's config, but I just wanted to check.

vaikas commented:
Thanks @haydentherapper I addressed your comments. I'll now convert the scaffolding to use this to make sure it works with the CTLog, since previously we were using the RSA keys for it.
https://github.com/sigstore/scaffolding/blob/main/cmd/ctlog/createctconfig/main.go#L114

haydentherapper reviewed (approved):
this looks great!

vaikas commented:
As I'm playing with this @haydentherapper and in particular with the tests about adding some new Fulcio to a pre-existing CTLog, we will have to deal with the fact that there may be RSA keys out there, so I'm trying to sort that out. Turns out, not everything is 'green-field' :)
